### PR TITLE
Add Encodec features for Codec toolkit

### PR DESCRIPTION
**Pull request by jctian98**
## What?

Add several Encodec features into the codec toolkit: https://arxiv.org/abs/2210.13438

- Multi-Scale STFT discriminator: done
- Skip discriminator training with a probability: done
- Loss Balancer: pending. 
@ftshijt

### CODE REVIEW

**Comment by ftshijt** [comment]
Since it is mixed with the current decoding and scoring PR, I will review it again after we first merge that one. Thanks for the in-time PR!

File: espnet2/gan_codec/encodec/encodec.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+            "average_by_layers": False,
+            "include_final_outputs": True,
+        },
+        use_mel_loss: bool = True,
+        mel_loss_params: Dict[str, Any] = {
+            "fs": 24000,
+            "n_fft": 1024,
+            "hop_length": 256,
+            "win_length": None,
+            "window": "hann",
```
Comment (ftshijt) [inline espnet2/gan_codec/encodec/encodec.py:85]: You might consider to follow the new setup for multi-scale mel loss?

File: espnet2/gan_codec/encodec/encodec.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        lambda_reconstruct: float = 1.0,
+        lambda_commit: float = 1.0,
+        lambda_adv: float = 1.0,
+        lambda_feat_match: float = 2.0,
+        lambda_mel: float = 45.0,
+        cache_generator_outputs: bool = False,
+        use_loss_balancer: bool = False,
+        balance_ema_decay: float = 0.99,
+    ):
+        # (Jinchuan) re-apply everything except the discriminator config.
```
Comment (ftshijt) [inline espnet2/gan_codec/encodec/encodec.py:102]: One TODO here, to add the docstring

Comment (ftshijt) [reply espnet2/gan_codec/encodec/encodec.py:102]: For the note consistency, please use:
```
# NOTE(jincuan): ...
```

Also for TODO

File: espnet2/gan_codec/shared/discriminator/msstft_conv.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+from torch import nn
+from torch.nn import functional as F
+from torch.nn.utils import spectral_norm, weight_norm
+
+CONV_NORMALIZATIONS = frozenset(
+    ["none", "weight_norm", "spectral_norm", "time_group_norm"]
+)
+
+
+def apply_parametrization_norm(module: nn.Module, norm: str = "none"):
```
Comment (ftshijt) [inline espnet2/gan_codec/shared/discriminator/msstft_conv.py:21]: For consistency, I think many of the modules are already in `shared`, could you please double check in case?

**Review by ftshijt — COMMENTED** [review]
Thanks for your contribution! In general, it looks great. Please check my minor comments~

**Comment by ftshijt** [comment]
Thanks for your contribution! Looks great to me.
